### PR TITLE
[ShopBundle] Index of products from given taxon

### PR DESCRIPTION
**Pull request by tuka217**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Related tickets  | -
| License         | MIT

### CODE REVIEW

File: features/product/viewing_products/viewing_products_form_specific_taxon.feature [hunk 1]
```diff
@@ -0,0 +1,33 @@
+@viewing_products
+Feature: Viewing products from specific taxon
```
Comment (TheMadeleine) [inline features/product/viewing_products/viewing_products_form_specific_taxon.feature]: **a** specific taxon

File: features/product/viewing_products/viewing_products_form_specific_taxon.feature [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    In order to see products from specific category
```
Comment (TheMadeleine) [inline features/product/viewing_products/viewing_products_form_specific_taxon.feature]: **a** specific category

File: features/product/viewing_products/viewing_products_form_specific_taxon.feature [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    As a Visitor
+    I want to be able to view products from selected category
```
Comment (TheMadeleine) [inline features/product/viewing_products/viewing_products_form_specific_taxon.feature]: **a** selected category

File: features/product/viewing_products/viewing_products_form_specific_taxon.feature [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        And the store has product "T-Shirt Banana" which is available in "Poland" channel
+        And this product belongs to "T-Shirts"
+        And this product belongs to "Funny"
+        And the store has product "Plastic Tomato" which is available in "Poland" channel
+        And this product belongs to "Funny"
+        And the store has product "T-Shirt Batman" which is available in "France" channel
+        And this product belongs to "T-Shirts"
+
+    @todo
+    Scenario: Viewing products from specific taxon
```
Comment (TheMadeleine) [inline features/product/viewing_products/viewing_products_form_specific_taxon.feature]: **a** specific taxon

File: features/product/viewing_products/viewing_products_form_specific_taxon.feature [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        Given I want to see products from specific category
```
Comment (TheMadeleine) [inline features/product/viewing_products/viewing_products_form_specific_taxon.feature]: **a** specific category

File: features/product/viewing_products/viewing_products_form_specific_taxon.feature [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        When I change my current channel to "Poland"
+        And I select "T-Shirts" category
+        Then I should see "T-Shirt Banana"
+        And I should not see "T-Shirt Batman"
+        And I should not see "Plastic Tomato"
+
+    @todo
+    Scenario: Viewing information about empty list of products from given category
+        Given I want to see products from specific category
```
Comment (TheMadeleine) [inline features/product/viewing_products/viewing_products_form_specific_taxon.feature]: Scenario: Viewing information about empty list of products from **a** given category
Given I want to see products from **a** specific category

File: features/product/viewing_products/viewing_products_form_specific_taxon.feature [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    I want to be able to view products from a selected taxon
+
+    Background:
+        Given the store has currency "Euro"
+        And the store operates on a channel named "Poland"
+        And this channel has currency "Euro"
+        And this channel host is "localhost"
+        And the store operates on another channel named "France"
+        And this channel has currency "Euro"
+        And this channel host is "localhost"
```
Comment (pjedrzejewski) [inline features/product/viewing_products/viewing_products_form_specific_taxon.feature]: Why?

File: features/product/viewing_products/viewing_products_form_specific_taxon.feature [hunk 8]
```diff
@@ -0,0 +1,36 @@
+@viewing_products
+Feature: Viewing products from a specific taxon
+    In order to see products from a specific taxon
```
Comment (pjedrzejewski) [inline features/product/viewing_products/viewing_products_form_specific_taxon.feature]: In order to browse products that interest me most

File: features/product/viewing_products/viewing_products_form_specific_taxon.feature [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    As a Visitor
+    I want to be able to view products from a selected taxon
```
Comment (pjedrzejewski) [inline features/product/viewing_products/viewing_products_form_specific_taxon.feature]: I want to be able to view products from a specific taxon

File: features/product/viewing_products/viewing_products_form_specific_taxon.feature [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        When I check list of products for taxon "T-Shirts"
+        Then I should see the product "T-Shirt Banana"
+        And I should not see the product "T-Shirt Batman"
+        And I should not see the product "Plastic Tomato"
+
+    @ui
+    Scenario: Viewing information about empty list of products from a given taxon
+        Given I want to see products in channel "France"
+        When I check list of products for taxon "Funny"
+        Then I should see information about empty list of products
```
Comment (pjedrzejewski) [inline features/product/viewing_products/viewing_products_form_specific_taxon.feature]: I think we should have 2 features in this PR:

1. For browsing products from taxon in a single channel store (here we only verify that we can view products from category)
2. For browsing products from taxon in a multi channel store (here we also verify that products are per channel)

That would simplify backgrounds of the 1st feature.

File: src/Sylius/Behat/Context/Setup/ChannelContext.php [hunk 11]
```diff
@@ -127,6 +128,26 @@ public function theChannelIsDisabled(ChannelInterface $channel)
...
+    {
+        $channel->addCurrency($currency);
+        $this->channelManager->flush();
+        $this->sharedStorage->set('channel', $channel);
+    }
+
+    /**
+     * @Given /^(this channel) host is "([^"]+)"$/
+     */
+    public function thisChannelHostIs(ChannelInterface $channel, $hostName)
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Setup/ChannelContext.php]: Why we need this? See https://github.com/Sylius/Sylius/blob/master/features/channel/theming/displaying_themed_channel_website.feature and its implementation in Behat's contexts to see how to switch between channels in test/dev env. :)

File: src/Sylius/Behat/Context/Ui/Shop/ProductContext.php [hunk 12]
```diff
@@ -111,4 +141,53 @@ public function iShouldSeeTheProductAttributeWithValue($attributeName, $Attribut
             sprintf('Product should have attribute %s with value %s, but it does not.', $attributeName, $AttributeValue)
         );
     }
+
+    /**
+     * @Given /^I want to see products in (channel "([^"]*)")$/
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Ui/Shop/ProductContext.php]: Ah, you've already figured this out - awesome! :) I'd just change this step to something like: ``I am browsing channel ...`` or something similar? To make it a generic step for switching channels. Maybe it deserves it's own Context to be reusable in all shop scenarios? We will have a lot of steps like that.

Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Ui/Shop/ProductContext.php]: CurrentChannelContext? Thoughts?

File: src/Sylius/Behat/Context/Ui/Shop/ProductContext.php [hunk 13]
```diff
@@ -111,4 +141,53 @@ public function iShouldSeeTheProductAttributeWithValue($attributeName, $Attribut
...
+     */
+    public function iWantToSeeProductsInChannel(ChannelInterface $channel)
+    {
+        $this->channelContextSetter->setChannel($channel);
+    }
+
+    /**
+     * @When /^I check list of products for (taxon "([^"]+)")$/
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Ui/Shop/ProductContext.php]: I browse products from taxon ...

File: src/Sylius/Behat/Context/Ui/Shop/ProductContext.php [hunk 14]
```diff
@@ -111,4 +141,53 @@ public function iShouldSeeTheProductAttributeWithValue($attributeName, $Attribut
...
+    public function iShouldNotSeeProduct($productName)
+    {
+        Assert::false(
+            $this->indexPage->isResourceOnPage($productName),
+            sprintf("The product %s should not appear on page, but it does.", $productName)
+        );
+    }
+
+    /**
+     * @Then I should see information about empty list of products
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Ui/Shop/ProductContext.php]: I should see empty list of products

File: src/Sylius/Behat/Page/Shop/Product/IndexPage.php [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+class IndexPage extends SymfonyPage implements IndexPageInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function isResourceOnPage($resourceName)
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: There will be no other resources, so I'd name this ``isProductInTheList`` or something similar.

File: src/Sylius/Behat/Page/Shop/Product/IndexPage.php [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+        return true;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getRouteName()
+    {
+        return 'sylius_shop_partial_product_index';
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: I think this should be ``sylius_shop_taxon_show`` and all this should go to ``Taxon\ShowPage``. Because we are indeed viewing a taxon and it's just a detail that we actually display a list of products there. For some taxons we might want to only display child categories and some recommended products and so on. So the "main concept" here is "Showing a Taxon" and product list is particular implementation.

To sum up: Merge these 2 pages in into taxon's show page.

File: src/Sylius/Behat/Resources/config/services/contexts/ui.xml [hunk 17]
```diff
@@ -293,6 +293,9 @@
 
         <service id="sylius.behat.context.ui.shop.product" class="%sylius.behat.context.ui.shop.product.class%" scope="scenario">
             <argument type="service" id="sylius.behat.page.shop.product.show" />
+            <argument type="service" id="sylius.behat.page.shop.product.index" />
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Resources/config/services/contexts/ui.xml]: According to my comments above, this page should be gone. If we ever add list of all products to Sylius (without taxon filtering) then this would be a correct page to handle this.

File: src/Sylius/Behat/Resources/config/services/pages/shop/product.xml [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+    <parameters>
+        <parameter key="sylius.behat.page.shop.product.show.class">Sylius\Behat\Page\Shop\Product\ShowPage</parameter>
+        <parameter key="sylius.behat.page.shop.product.index.class">Sylius\Behat\Page\Shop\Product\IndexPage</parameter>
+    </parameters>
+
+    <services>
+        <service id="sylius.behat.page.shop.product.show" class="%sylius.behat.page.shop.product.show.class%" parent="sylius.behat.symfony_page" scope="scenario" public="false" />
+        <service id="sylius.behat.page.shop.product.index" class="%sylius.behat.page.shop.product.index.class%" parent="sylius.behat.symfony_page" scope="scenario" public="false" />
+    </services>
+</container>
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Resources/config/services/pages/shop/product.xml]: Please keep this file when removing the index page of product, this separation is good.

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php [hunk 19]
```diff
@@ -207,6 +207,23 @@ public function findOneByIdAndChannel($id, ChannelInterface $channel = null)
     /**
      * {@inheritdoc}
      */
+    public function findByTaxonCodeAndChannel($code, ChannelInterface $channel)
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php]: We should also consider getting only enabled products.

``findEnabledByTaxonCodeAndChannel``?

File: src/Sylius/Bundle/ShopBundle/Resources/config/routing.yml [hunk 20]
```diff
@@ -23,3 +23,10 @@ sylius_shop_security:
 sylius_shop_checkout:
     prefix: /checkout
     resource: "@SyliusShopBundle/Resources/config/routing/checkout.yml"
+
+sylius_shop_taxon:
+    resource: "@SyliusShopBundle/Resources/config/routing/taxon.yml"
+
+sylius_shop_partial_product:
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ShopBundle/Resources/config/routing.yml]: All partials should go to separate file, just like in admin. Also you can have a look at #5138.

File: src/Sylius/Bundle/ShopBundle/Resources/config/routing.yml [hunk 21]
```diff
@@ -23,3 +23,10 @@ sylius_shop_security:
...
+    resource: "@SyliusShopBundle/Resources/config/routing/partial/product.yml"
+    prefix: partial
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ShopBundle/Resources/config/routing.yml]: /partial

File: src/Sylius/Bundle/ShopBundle/Resources/config/routing/taxon.yml [hunk 22]
```diff
@@ -0,0 +1,15 @@
+# This file is part of the Sylius package.
+# (c) Paweł Jędrzejewski
+
+sylius_shop_taxon_show:
+    path: taxons/{permalink}
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ShopBundle/Resources/config/routing/taxon.yml]: Always add ``/`` at the beginning of path.

File: src/Sylius/Behat/Context/Setup/ChannelContext.php (hunk 11 above)
Comment (pamil) [inline src/Sylius/Behat/Context/Setup/ChannelContext.php]: $hostName -> $hostname

File: src/Sylius/Behat/Context/Setup/ChannelContext.php [hunk 23]
```diff
@@ -127,6 +128,26 @@ public function theChannelIsDisabled(ChannelInterface $channel)
...
+    {
+        $channel->setHostname($hostName);
+        $this->channelManager->flush();
+        $this->sharedStorage->set('channel', $channel);
```
Comment (pamil) [inline src/Sylius/Behat/Context/Setup/ChannelContext.php]: It's already in shared storage.

File: features/product/viewing_products/viewing_products_form_specific_taxon.feature [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+        And this product belongs to "Funny"
+        And the store has a product "Plastic Tomato" available in "Poland" channel
+        And this product belongs to "Funny"
+        And the store has a product "T-Shirt Batman" available in "France" channel
+        And this product belongs to "T-Shirts"
+
+    @ui
+    Scenario: Viewing products from a specific taxon
+        Given I want to see products in channel "Poland"
+        When I check list of products for taxon "T-Shirts"
```
Comment (michalmarcinkowski) [inline features/product/viewing_products/viewing_products_form_specific_taxon.feature]: When I **browse** products for taxon "T-Shirts"

File: src/Sylius/Behat/Page/Shop/Product/IndexPage.php [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+        return false;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isEmpty()
+    {
+        $isEmpty = strpos($this->getDocument()->find('css', '.message')->getText(), 'There are no products to display');
+        if (false === $isEmpty ) {
```
Comment (pamil) [inline src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: CS :)

Comment (pamil) [inline src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: Anyway, this function can be a oneliner:

```
return false !== strpos($this->getDocument()->find('css', '.message')->getText(), 'There are no products to display');
```

File: src/Sylius/Behat/Context/Ui/Shop/ProductContext.php (hunk 12 above)
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Ui/Shop/ProductContext.php]: `CurrentChannelContext` sounds good! And the step e.g. `Given I am browsing "([^"]*)" channel`.

File: src/Sylius/Bundle/ShopBundle/Resources/views/Product/_singleProduct.html.twig [hunk 26]
```diff
@@ -0,0 +1,13 @@
+{% from 'SyliusShopBundle:Product:macros.html.twig' import sylius_display_price %}
+
+<div class="content">
+    <div class="header">{{ product.name|truncate(18) }}</div>
+</div>
+<div class="extra content">
+    <span class="right floated">
+        <button class="ui primary button">Add to cart</button>
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Product/_singleProduct.html.twig]: Should be translatable.

File: src/Sylius/Behat/Context/Setup/ChannelContext.php (hunk 11 above)
Comment (tuka217) [inline src/Sylius/Behat/Context/Setup/ChannelContext.php]: Yup, but before I couldn't  open my taxon show page with configuration like this from channel feature. Now is ok and I can remove this steps.

File: src/Sylius/Bundle/ShopBundle/Resources/views/Product/_singleProduct.html.twig [hunk 27]
```diff
@@ -0,0 +1,13 @@
...
+    </span>
+    <span>
+        {{ sylius_display_price(product.firstVariant) }}
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Product/_singleProduct.html.twig]: We should add `if (product.firstVariant is not null)` here, because we allow for products without variants.

File: src/Sylius/Bundle/ShopBundle/Resources/views/Product/macros.html.twig [hunk 28]
```diff
@@ -4,3 +4,23 @@
     {% endif %}
     <span> {{ sylius_calculate_price(variant)|sylius_price }} </span>
 {% endmacro %}
+
+{% macro list(products, limit = 2) %}
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Product/macros.html.twig]: Is there any use case we would want to have other value than `2` of the `limit` variable?

File: src/Sylius/Behat/Context/Ui/Shop/ProductContext.php (hunk 12 above)
Comment (tuka217) [inline src/Sylius/Behat/Context/Ui/Shop/ProductContext.php]: Hmm.. new context for one step? Unless there will be more than one then ok.

Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Ui/Shop/ProductContext.php]: This step will be used in pretty much every scenario that considers more than 1 channel. So it is better to have single step context than duplicated step in all contexts.

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php (hunk 19 above)
Comment (tuka217) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php]: A product is diabled permanently? If not, then why customer cannot see product avaiable for them but not exactly in this moment?

Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php]: Yes, it is disabled permanently. ``enabled`` flag indicates that. Being out-of-stock, current unavailable, discontinued are separate features and will be handled differently.

File: src/Sylius/Bundle/ShopBundle/Resources/views/Product/macros.html.twig (hunk 28 above)
Comment (tuka217) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Product/macros.html.twig]: If _singleProduct element on page will have small portion of information and little pictures then you may want to put more than 2 element in one row. Everything depends on  _singleProducts final form, becouse now it has only the most necessary fields.

**Comment by pjedrzejewski** [comment]
Nice work, thank you Ania! 👍
